packages/image_picker/example/android/app/src/test/java/io/flutter/plugins/imagepicker/ImagePickerPluginTest.java

```diff
@@ -0,0 +1,107 @@
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+public class ImagePickerPluginTest {
+  private static final int SOURCE_CAMERA = 0;
+  private static final int SOURCE_GALLERY = 1;
+
+  @Rule public ExpectedException exception = ExpectedException.none();
+
+  @Mock PluginRegistry.Registrar mockRegistrar;
+  @Mock Activity mockActivity;
+  @Mock ImagePickerDelegate mockImagePickerDelegate;
+  @Mock MethodChannel.Result mockResult;
+
+  ImagePickerPlugin plugin;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+
+    plugin = new ImagePickerPlugin(mockRegistrar, mockImagePickerDelegate);
+  }
+
+  @Test
+  public void onMethodCall_WhenActivityIsNull_FinishesWithForegroundActivityRequiredError() {
+    when(mockRegistrar.activity()).thenReturn(null);
+    MethodCall call = buildMethodCall(SOURCE_GALLERY);
+
+    plugin.onMethodCall(call, mockResult);
+
+    verify(mockResult)
+        .error("no_activity", "image_picker plugin requires a foreground activity.", null);
+    verifyZeroInteractions(mockImagePickerDelegate);
+  }
+
+  @Test
+  public void onMethodCall_WhenCalledWithUnknownMethod_ThrowsException() {
+    when(mockRegistrar.activity()).thenReturn(mockActivity);
+    exception.expect(IllegalArgumentException.class);
+    exception.expectMessage("Unknown method test");
+
+    plugin.onMethodCall(new MethodCall("test", null), mockResult);
+
+    verifyZeroInteractions(mockImagePickerDelegate);
+    verifyZeroInteractions(mockResult);
+  }
+
+  @Test
+  public void onMethodCall_WhenCalledWithUnknownImageSource_ThrowsException() {
+    when(mockRegistrar.activity()).thenReturn(mockActivity);
+    exception.expect(IllegalArgumentException.class);
+    exception.expectMessage("Invalid image source: -1");
+
+    plugin.onMethodCall(buildMethodCall(-1), mockResult);
+
+    verifyZeroInteractions(mockImagePickerDelegate);
+    verifyZeroInteractions(mockResult);
+  }
+
+  @Test
+  public void onMethodCall_WhenSourceIsGallery_InvokesChooseImageFromGallery() {
+    when(mockRegistrar.activity()).thenReturn(mockActivity);
+    MethodCall call = buildMethodCall(SOURCE_GALLERY);
+
+    plugin.onMethodCall(call, mockResult);
+
+    verify(mockImagePickerDelegate).chooseImageFromGallery(call, mockResult);
+    verifyZeroInteractions(mockResult);
+  }
+
+  @Test
+  public void onMethodCall_WhenSourceIsCamera_InvokesTakeImageWithCamera() {
+    when(mockRegistrar.activity()).thenReturn(mockActivity);
+    MethodCall call = buildMethodCall(SOURCE_CAMERA);
+
+    plugin.onMethodCall(call, mockResult);
+
+    verify(mockImagePickerDelegate).takeImageWithCamera(call, mockResult);
+    verifyZeroInteractions(mockResult);
+  }
+
+  private MethodCall buildMethodCall(final int source) {
+    final HashMap<String, Object> arguments =
+        new HashMap<String, Object>() {
+          {
+            put("source", source);
+          }
+        };
```

Comment:
Can be shortened to
```java
final Map<String, Object> arguments = new HashMap<>();
arguments.put("source", source);
```

Reply:
Sure! I usually use that syntax when I have 3 or more things to `put` there, but for this case it's a little redundant.

Reply:
:-) Map and List literals ftw. You get a `HashMap` subclass and are forced to repeat the type args. I usually add classes `CreateList`, `CreateMap`, `CreateSet` to any non-trivial Java project with static `.of(...)` methods that create, populate and return small collections. It makes unit tests so much more concise.